File: apps/web/src/lib/mcp/device-mcp-servers.ts
// EXP-891: MCP servers a MACHINE connects on its own runs — per device and
// per member, beside the team registry (EXP-792, `lib/mcp-servers.ts`).
//
// The decision (EXP-988 open decision 1): a NEW server-only table,
// `device_mcp_servers`, not a widening of `mcp_servers`. The team registry
// is a shared, owner-authored catalogue with an auth kind, a per-device
// readiness matrix and an OAuth flow behind every row; a device row is the
// opposite — private to one machine and its member, written by the DEVICE
// from what the local claude/codex config already holds (`claude mcp add …`,
// `[mcp_servers.*]` in codex's config.toml), with no credential position at
// all: whatever auth the agent already keeps for that server stays where it
// is. Rows are edited only on the IDE / CLI (import, add, enable, forget);
// the web shows them read-only, grouped per device. Every enabled row is
// appended to every run the device starts (`coding::device_mcp_servers`),
// so there is nothing to pick on the Agent page.
import type { McpTransport } from "@exp/db-schema/domain"

export const deviceMcpServerSources = [`detected`, `manual`] as const
export type DeviceMcpServerSource = (typeof deviceMcpServerSources)[number]

/** The local agent configs a `detected` row can come from. */
export const deviceMcpServerAgents = [`claude`, `codex`] as const
export type DeviceMcpServerAgent = (typeof deviceMcpServerAgents)[number]

export interface DeviceMcpServer {
  id: string
  /** `devices.device_id` (the machine's stable id, not the row id). */
  deviceId: string
  /** The team member who owns the device. */
  userId: string
  name: string
  transport: McpTransport
  /** `http` transport. */
  url?: string | null
  /** `stdio` transport. */
  command?: string | null
  /** `detected` = imported from the local agent config; `manual` = typed. */
  source: DeviceMcpServerSource
  enabled: boolean
}

/** One row as `deviceMcpServers.list` returns it: the contract shape plus
 * what the web needs to group and caption it. */
export interface DeviceMcpServerListRow extends DeviceMcpServer {
  /** The `devices` row id. */
  deviceRowId: string
  deviceLabel: string
  /** The device owner's display name — the caller's own name on own rows. */
  ownerName: string
  /** `stdio` arguments. */
  args: string[]
  /** The local config a detected row was imported from; null when typed. */
  agent: DeviceMcpServerAgent | null
  updatedAt: string
}

/** The per-machine set the device pushes with `deviceMcpServers.sync` —
 * the whole list, replacing what the server held for that device. */
export interface DeviceMcpServerInput {
  name: string
  transport: McpTransport
  url?: string | null
  command?: string | null
  args?: string[]
  source: DeviceMcpServerSource
  agent?: DeviceMcpServerAgent | null
  enabled: boolean
}

/** Rows one device may hold (a hand-written config never lists more). */
export const MAX_DEVICE_MCP_SERVERS = 64
export const MAX_DEVICE_MCP_SERVER_NAME = 64

/** The web hand-off line: where a row gets added, imported or switched. */
export const DEVICE_MCP_SETUP_HINT = `Servers each machine connects on its own runs. Import or add them in the desktop app under Settings › MCP servers, or with the CLI: exponential mcp import, exponential mcp add.`

/** `https://mcp.linear.app/mcp` for http, the command line for stdio. */
export function deviceMcpServerTarget(
  row: Pick<DeviceMcpServer, `transport` | `url` | `command`> & {
    args?: string[] | null
  }
): string {
  if (row.transport === `http`) return row.url ?? ``
  return [row.command, ...(row.args ?? [])].filter(Boolean).join(` `)
}

/** The provenance chip: `Detected from Claude Code`, `Detected from Codex`,
 * `Added on the device`. */
export function deviceMcpServerSourceLabel(
  row: Pick<DeviceMcpServer, `source`> & {
    agent?: DeviceMcpServerAgent | string | null
  }
): string {
  if (row.source === `detected`) {
    if (row.agent === `claude`) return `Detected from Claude Code`
    if (row.agent === `codex`) return `Detected from Codex`
    return `Detected on the device`
  }
  return `Added on the device`
}

export interface DeviceMcpServerGroup {
  deviceId: string
  deviceRowId: string
  deviceLabel: string
  userId: string
  ownerName: string
  /** Name order, case-insensitive. */
  servers: DeviceMcpServerListRow[]
}

/** Rows grouped per device (the settings pane's one band per machine):
 * the caller's own machines first, then teammates' shared ones, each set
 * alphabetical by label; rows inside a group by name. */
export function groupDeviceMcpServers(
  rows: readonly DeviceMcpServerListRow[],
  currentUserId: string
): DeviceMcpServerGroup[] {
  const byDevice = new Map<string, DeviceMcpServerGroup>()
  for (const row of rows) {
    let group = byDevice.get(row.deviceRowId)
    if (!group) {
      group = {
        deviceId: row.deviceId,
        deviceRowId: row.deviceRowId,
        deviceLabel: row.deviceLabel,
        userId: row.userId,
        ownerName: row.ownerName,
        servers: [],
      }
      byDevice.set(row.deviceRowId, group)
    }
    group.servers.push(row)
  }
  const byName = (a: { name: string }, b: { name: string }) =>
    a.name.localeCompare(b.name, undefined, { sensitivity: `base` })
  const groups = [...byDevice.values()]
  for (const group of groups) group.servers.sort(byName)
  groups.sort((a, b) => {
    const mineA = a.userId === currentUserId ? 0 : 1
    const mineB = b.userId === currentUserId ? 0 : 1
    if (mineA !== mineB) return mineA - mineB
    return (a.deviceLabel || a.deviceId).localeCompare(
      b.deviceLabel || b.deviceId,
      undefined,
      { sensitivity: `base` }
    )
  })
  return groups
}

/** The row rules `deviceMcpServers.sync` enforces, shared with the input
 * schema so a refusal names the field: an http row needs a URL (https, or
 * http on the machine's own loopback — a local dev MCP never leaves it), a
 * stdio row a command, and no row may fold to the launcher's reserved
 * `exponential` config key. Returns null for a valid row. */
export function validateDeviceMcpServerInput(
  input: DeviceMcpServerInput
): string | null {
  const name = input.name.trim()
  if (name.length === 0) return `A server needs a name`
  if (name.length > MAX_DEVICE_MCP_SERVER_NAME) {
    return `Server names are at most ${MAX_DEVICE_MCP_SERVER_NAME} characters`
  }
  if (configKey(name) === `exponential`) return `The name ${name} is reserved`
  if (input.transport === `http`) {
    const url = (input.url ?? ``).trim()
    if (url.length === 0) return `${name}: an http server needs a URL`
    let parsed: URL
    try {
      parsed = new URL(url)
    } catch {
      return `${name}: that URL does not parse`
    }
    const loopback =
      parsed.hostname === `localhost` || parsed.hostname === `127.0.0.1`
    if (
      parsed.protocol !== `https:` &&
      !(parsed.protocol === `http:` && loopback)
    ) {
      return `${name}: the URL must be https:// (http:// only for localhost)`
    }
  } else if ((input.command ?? ``).trim().length === 0) {
    return `${name}: a stdio server needs a command`
  }
  return null
}

/** The config key a name becomes in the rendered agent config (the desktop
 * `McpServerWire::config_key` twin): lowercase ASCII alphanumerics, the rest
 * folded to `_`, never empty. Two rows folding to one key would shadow each
 * other, so `sync` refuses the second. */
export function configKey(name: string): string {
  const key = name
    .trim()
    .split(``)
    .map((c) => (/[A-Za-z0-9]/.test(c) ? c.toLowerCase() : `_`))
    .join(``)
  return key.length === 0 ? `server` : key
}

File: apps/desktop/crates/api/src/device_mcp_servers.rs
//! EXP-891: typed `deviceMcpServers.*` tRPC helpers — the device side of
//! PER-DEVICE MCP servers (beside the team registry in
//! [`crate::mcp_servers`]). The machine is the writer: `sync` replaces the
//! server's copy of THIS device's whole set with what the local store holds
//! ([`coding::device_mcp_servers`]). Config only — a url or a command +
//! args — never a header value, a token or an env value.
//!
//! Shapes mirror `apps/web/src/lib/trpc/device-mcp-servers.ts`.

use serde::{Deserialize, Serialize};

use crate::error::ApiError;
use crate::trpc::TrpcClient;

/// One row of the machine's set as `deviceMcpServers.sync` takes it.
#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct DeviceMcpServerInput {
    pub name: String,
    /// `http` | `stdio`.
    pub transport: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub url: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub command: Option<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub args: Vec<String>,
    /// `detected` | `manual`.
    pub source: String,
    /// `claude` | `codex` for a detected row; absent for a typed one.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub agent: Option<String>,
    pub enabled: bool,
}

/// `deviceMcpServers.sync` — replace this device's set. Returns how many rows
/// the server now holds for it.
pub fn sync(
    trpc: &TrpcClient,
    device_id: &str,
    servers: &[DeviceMcpServerInput],
) -> Result<usize, ApiError> {
    #[derive(Serialize)]
    #[serde(rename_all = "camelCase")]
    struct Input<'a> {
        device_id: &'a str,
        servers: &'a [DeviceMcpServerInput],
    }
    #[derive(Deserialize)]
    struct Ok {
        #[serde(default)]
        count: usize,
    }
    let ok: Ok = trpc.mutation("deviceMcpServers.sync", &Input { device_id, servers })?;
    Ok(ok.count)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::trpc::tests::one_shot_server;
    use crate::StaticToken;
    use std::sync::Arc;

    fn client(base: &str) -> TrpcClient {
        TrpcClient::new(base, Arc::new(StaticToken("tok".to_string())))
    }

    /// The wire is `{deviceId, servers: [...]}` with absent optionals left
    /// off — the router's zod schema takes `nullish` there, and an empty
    /// `args` is the stdio default.
    #[test]
    fn sync_posts_the_set_under_the_device_id() {
        let (base, rx) = one_shot_server(200, r#"{"result":{"data":{"ok":true,"count":2}}}"#);
        let count = sync(
            &client(&base),
            "dev-1",
            &[
                DeviceMcpServerInput {
                    name: "linear".into(),
                    transport: "http".into(),
                    url: Some("https://mcp.linear.app/mcp".into()),
                    source: "detected".into(),
                    agent: Some("claude".into()),
                    enabled: true,
                    ..Default::default()
                },
                DeviceMcpServerInput {
                    name: "acme".into(),
                    transport: "stdio".into(),
                    command: Some("npx".into()),
                    args: vec!["-y".into(), "@acme/mcp".into()],
                    source: "manual".into(),
                    enabled: false,
                    ..Default::default()
                },
            ],
        )
        .expect("ok");
        assert_eq!(count, 2);
        let request = rx.recv().unwrap();
        assert!(request.contains("POST /api/trpc/deviceMcpServers.sync"));
        assert!(request.contains(
            r#"{"deviceId":"dev-1","servers":[{"name":"linear","transport":"http","url":"https://mcp.linear.app/mcp","source":"detected","agent":"claude","enabled":true},{"name":"acme","transport":"stdio","command":"npx","args":["-y","@acme/mcp"],"source":"manual","enabled":false}]}"#
        ), "{request}");
    }

    #[test]
    fn sync_of_an_empty_set_clears_the_device() {
        let (base, rx) = one_shot_server(200, r#"{"result":{"data":{"ok":true,"count":0}}}"#);
        assert_eq!(sync(&client(&base), "dev-1", &[]).expect("ok"), 0);
        let request = rx.recv().unwrap();
        assert!(request.contains(r#"{"deviceId":"dev-1","servers":[]}"#));
    }
}
